fleetspeak/src/osquery/plugin/distributed_test.go

```diff
@@ -0,0 +1,173 @@
+			},
+			inResults: []osquery.ExtensionPluginRequest{
+				{"action": "writeResults", "results": `{"queries":{"multi1":[{"name":"client1"}], "multi2":[{"name":"client2"}], "multi3":[{"name":"client3"}], "multi4":[{"name":"client4"}]}, "statuses":{"multi1":"0","multi2":"0","multi3":"0","multi4":"0"}}`},
+			},
+			wantRows: map[string]*ospb.Rows{
+				"multi1": &ospb.Rows{Rows: []*ospb.Row{{Row: map[string]string{"name": "client1"}}}},
+				"multi2": &ospb.Rows{Rows: []*ospb.Row{{Row: map[string]string{"name": "client2"}}}},
+				"multi3": &ospb.Rows{Rows: []*ospb.Row{{Row: map[string]string{"name": "client3"}}}},
+				"multi4": &ospb.Rows{Rows: []*ospb.Row{{Row: map[string]string{"name": "client4"}}}},
+			},
+		},
+		{
+			name: "multi-results",
+			inQueries: []*ospb.Queries{{
+				Queries: map[string]string{"basic": "SELECT * FROM foobar"},
+			}},
+			wantGet: osquery.ExtensionPluginResponse{
+				{"results": `{"queries":{"basic":"SELECT * FROM foobar"}}`},
+			},
+			inResults: []osquery.ExtensionPluginRequest{
+				{"action": "writeResults", "results": `{"queries":{"basic":[{"iso_8601":"2017-07-10T22:08:40Z"}, {"iso_8601":"2017-07-10T22:08:41Z"}, {"iso_8601":"2017-07-10T22:08:42Z"}]}, "statuses":{"basic":"0"}}`},
+			},
+			wantRows: map[string]*ospb.Rows{"basic": &ospb.Rows{Rows: []*ospb.Row{
+				{Row: map[string]string{"iso_8601": "2017-07-10T22:08:40Z"}},
+				{Row: map[string]string{"iso_8601": "2017-07-10T22:08:41Z"}},
+				{Row: map[string]string{"iso_8601": "2017-07-10T22:08:42Z"}},
+			}}},
+		},
+	} {
+		serviceName := fmt.Sprintf("TestService-%s", tc.name)
+		for _, q := range tc.inQueries {
+			data, err := ptypes.MarshalAny(q)
+			if err != nil {
+				t.Errorf("%s: Failed to marshal ospb.Queries: %v", tc.name, err)
+			}
+			toOSQuery <- &fspb.Message{
+				Source:      &fspb.Address{ServiceName: serviceName},
+				MessageType: "Queries",
+				Data:        data,
+			}
+		}
+		res := dist.Call(context.Background(), osquery.ExtensionPluginRequest{"action": "getQueries"})
+		if (*res.Status != osquery.ExtensionStatus{Code: 0, Message: "OK"}) {
+			t.Errorf("%s: Unexpected ExtensionStatus, wanted OK, got: %v", tc.name, res.Status)
+			continue
+		}
+		if !reflect.DeepEqual(res.Response, tc.wantGet) {
+			t.Errorf("%s: Unexpected Response attributed, wanted %v, got %v.", tc.name, tc.wantGet, res.Response)
+			continue
+		}
+		for _, r := range tc.inResults {
+			res := dist.Call(context.Background(), r)
+			if (*res.Status != osquery.ExtensionStatus{Code: 0, Message: "OK"}) {
+				t.Errorf("%s: Unexpected ExtensionStatus, wanted OK, got: %v", tc.name, res.Status)
+				continue
+			}
+		}
+	L:
+		for {
+			var msg *fspb.Message
+			select {
+			case msg = <-fromOSQuery:
+			default:
+				break L
+			}
+
+			if !proto.Equal(msg.Destination, &fspb.Address{ServiceName: serviceName}) {
+				t.Errorf("%s: Unexpected Destination, wanted %v, got %v", tc.name, msg.Destination, &fspb.Address{ServiceName: serviceName})
```

Comment:
Args for 'want' and 'got' are switched?

Reply:
Oops, clicked merge before I saw this, will send a fixup PR.